### PR TITLE
First Draft

### PR DESCRIPTION
**Pull request by knutwalker**

### CODE REVIEW

**Comment by jexp** [comment]
* Please rename all vertex -> node and edge -> relationship
* please squash the commits into a few sensible ones (API, infrastructure, tests, procedures)
 (or just one if it's too much effort this time)

File: doc/index.adoc [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+This implementations utilizes an int-matrix for storing connections between vertices.
+It has a high memory consumption but performs basic calculations on the graph
+around 3 times faster then the memory efficient implementation. Furthermore the
+number of edges per node is limited only to the maximum array size of the VM and
+connections can be added in arbitrary order.
+
+LightGraph::
+
+This implementation takes 3 times less heap due to a more intelligent
+memory layout. The drawback is the higher evaluation time.
```
Comment (jexp) [inline doc/index.adoc]: can we figure out why the light-graph is slower, from what I saw it shouldn't be.

File: pom.xml [hunk 2]
```diff
@@ -1,93 +1,123 @@
...
+            <dependency>
+                <groupId>org.neo4j</groupId>
+                <artifactId>neo4j</artifactId>
+                <version>${neo4j.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.neo4j</groupId>
+                <artifactId>neo4j-kernel</artifactId>
+                <version>${neo4j.version}</version>
+                <type>test-jar</type>
```
Comment (jexp) [inline pom.xml]: scope test

File: pom.xml [hunk 3]
```diff
@@ -1,93 +1,123 @@
...
-        <groupId>org.neo4j</groupId>
-        <artifactId>procedure-compiler</artifactId>
-        <version>${neo4j.version}</version>
-        <scope>compile</scope>
-        <optional>true</optional>
-    </dependency>
+            <dependency>
+                <groupId>org.neo4j</groupId>
+                <artifactId>neo4j</artifactId>
+                <version>${neo4j.version}</version>
```
Comment (jexp) [inline pom.xml]: scope provided

File: pom.xml [hunk 4]
```diff
@@ -1,93 +1,123 @@
...
+                <groupId>org.neo4j</groupId>
+                <artifactId>neo4j-kernel</artifactId>
+                <version>${neo4j.version}</version>
+                <type>test-jar</type>
+            </dependency>
+            <dependency>
+                <groupId>org.neo4j</groupId>
+                <artifactId>neo4j-io</artifactId>
+                <version>${neo4j.version}</version>
+                <type>test-jar</type>
```
Comment (jexp) [inline pom.xml]: scope test

File: tests/pom.xml [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>graph-algorithms-parent</artifactId>
+        <groupId>org.neo4j</groupId>
+        <version>3.1.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>graph-algrotihms-tests</artifactId>
```
Comment (jexp) [inline tests/pom.xml]: typo

File: tests/pom.xml [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        <graph-algorithms-project.version>3.1.0-SNAPSHOT</graph-algorithms-project.version>
+    </properties>
+
+    <dependencies>
+
+        <dependency>
+            <groupId>org.neo4j</groupId>
+            <artifactId>neo4j-io</artifactId>
+            <version>${neo4j.version}</version>
+            <scope>compile</scope>
```
Comment (jexp) [inline tests/pom.xml]: scope test

File: tests/pom.xml [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+            <groupId>org.neo4j</groupId>
+            <artifactId>neo4j-kernel</artifactId>
+            <version>3.1.1</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.neo4j</groupId>
+            <artifactId>neo4j-kernel</artifactId>
+            <version>3.1.1</version>
+            <scope>compile</scope>
```
Comment (jexp) [inline tests/pom.xml]: scope test

File: tests/pom.xml [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+            <artifactId>procedure-compiler</artifactId>
+            <version>${neo4j.version}</version>
+            <scope>compile</scope>
+            <optional>true</optional>
+        </dependency>
+
+        <dependency>
+            <groupId>org.neo4j</groupId>
+            <artifactId>neo4j-cypher</artifactId>
+            <version>${neo4j.version}</version>
```
Comment (jexp) [inline tests/pom.xml]: just neo4j dependency should be fine

File: tests/pom.xml [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest-library</artifactId>
+            <version>1.3</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.neo4j</groupId>
+            <artifactId>neo4j-kernel</artifactId>
+            <version>3.1.1</version>
```
Comment (jexp) [inline tests/pom.xml]: scope provided

File: tests/pom.xml [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+        <dependency>
+            <groupId>org.neo4j</groupId>
+            <artifactId>graph-algrotihms-core</artifactId>
+            <version>${graph-algorithms-project.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.neo4j</groupId>
+            <artifactId>neo4j-primitive-collections</artifactId>
+            <version>${neo4j.version}</version>
```
Comment (jexp) [inline tests/pom.xml]: scope provided?

File: tests/pom.xml [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>${junit.version}</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+            <version>${slf4j.version}</version>
```
Comment (jexp) [inline tests/pom.xml]: neo4j comes with logging that can be tied in

File: tests/pom.xml [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+                    <execution>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>shade</goal>
+                        </goals>
+                        <configuration>
+                            <finalName>test-assembly</finalName>
+                            <transformers>
+                                <transformer implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
+                                    <mainClass>de.avgl.graphtest.tests.extra.LargeishPageRanker</mainClass>
```
Comment (jexp) [inline tests/pom.xml]: classname?

File: tests/src/main/java/de/avgl/graphtest/tests/extra/GcLogger.java [hunk 13]
```diff
@@ -0,0 +1,197 @@
+package de.avgl.graphtest.tests.extra;
```
Comment (jexp) [inline tests/src/main/java/de/avgl/graphtest/tests/extra/GcLogger.java]: wrong package name?

File: tests/src/main/java/de/avgl/graphtest/tests/extra/GcRunner.java [hunk 14]
```diff
@@ -0,0 +1,5 @@
+package de.avgl.graphtest.tests.extra;
```
Comment (jexp) [inline tests/src/main/java/de/avgl/graphtest/tests/extra/GcRunner.java]: package name

File: tests/src/main/java/de/avgl/graphtest/tests/extra/LargeishPageRanker.java [hunk 15]
```diff
@@ -0,0 +1,245 @@
+package de.avgl.graphtest.tests.extra;
```
Comment (jexp) [inline tests/src/main/java/de/avgl/graphtest/tests/extra/LargeishPageRanker.java]: package name

File: tests/src/test/java/de/avgl/graphtest/tests/GraphComparisionTest.java [hunk 16]
```diff
@@ -0,0 +1,82 @@
+package de.avgl.graphtest.tests;
```
Comment (jexp) [inline tests/src/test/java/de/avgl/graphtest/tests/GraphComparisionTest.java]: package

File: tests/src/test/java/de/avgl/graphtest/tests/GraphComparisionTest.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+ * @author mknobloch
+ */
+@Ignore("depends on local data")
+public class GraphComparisionTest {
+
+    private static final Logger LOGGER = getLogger(GraphComparisionTest.class);
+
+    private GraphDatabaseService db;
+
+    private static final Path GRAPH_PATH = Paths.get("/Users/mknobloch/Documents/graph/");
```
Comment (jexp) [inline tests/src/test/java/de/avgl/graphtest/tests/GraphComparisionTest.java]: we can put the data in s3 or gdrive

File: tests/src/test/java/de/avgl/graphtest/tests/IntMinPriorityQueueTest.java [hunk 18]
```diff
@@ -0,0 +1,119 @@
+package de.avgl.graphtest.tests;
```
Comment (jexp) [inline tests/src/test/java/de/avgl/graphtest/tests/IntMinPriorityQueueTest.java]: package

File: tests/src/test/java/de/avgl/graphtest/tests/GraphComparisionTest.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+    @Test
+    public void testLight() throws Exception {
+        final Result result = db.execute("CALL graphalgo.pagerank.light({iterations:20, dampingFactor:0.85})");
+        assertEquals(3697509, result.stream().count());
+    }
+
+    @Test
+    public void testAlgo() throws Exception {
+        final Result result = db.execute("CALL algo.algo.pageRankWithCypher({iterations:20, write:false})");
+        assertTrue(result.hasNext());
```
Comment (jexp) [inline tests/src/test/java/de/avgl/graphtest/tests/GraphComparisionTest.java]: better test?

**Comment by jexp** [comment]
It would also be good to have some more tests of the Graph/Loader implementations.
Esp. also as needed in multi-threaded scenarios.

File: tests/src/test/java/de/avgl/graphtest/tests/IntMinPriorityQueueTest.java (hunk 18 above)
Comment (mknblch) [inline tests/src/test/java/de/avgl/graphtest/tests/IntMinPriorityQueueTest.java]: done

File: tests/src/test/java/de/avgl/graphtest/tests/GraphComparisionTest.java (hunk 19 above)
Comment (mknblch) [inline tests/src/test/java/de/avgl/graphtest/tests/GraphComparisionTest.java]: done

File: tests/src/test/java/de/avgl/graphtest/tests/GraphComparisionTest.java (hunk 16 above)
Comment (mknblch) [inline tests/src/test/java/de/avgl/graphtest/tests/GraphComparisionTest.java]: done

File: tests/src/main/java/de/avgl/graphtest/tests/extra/LargeishPageRanker.java (hunk 15 above)
Comment (mknblch) [inline tests/src/main/java/de/avgl/graphtest/tests/extra/LargeishPageRanker.java]: done

File: tests/src/main/java/de/avgl/graphtest/tests/extra/GcRunner.java (hunk 14 above)
Comment (mknblch) [inline tests/src/main/java/de/avgl/graphtest/tests/extra/GcRunner.java]: done

File: tests/src/main/java/de/avgl/graphtest/tests/extra/GcLogger.java (hunk 13 above)
Comment (mknblch) [inline tests/src/main/java/de/avgl/graphtest/tests/extra/GcLogger.java]: done

File: tests/pom.xml (hunk 12 above)
Comment (mknblch) [inline tests/pom.xml]: done

File: tests/pom.xml (hunk 11 above)
Comment (mknblch) [inline tests/pom.xml]: done

File: tests/pom.xml (hunk 10 above)
Comment (mknblch) [inline tests/pom.xml]: done

File: tests/pom.xml (hunk 8 above)
Comment (mknblch) [inline tests/pom.xml]: done

File: tests/pom.xml (hunk 9 above)
Comment (mknblch) [inline tests/pom.xml]: done

File: tests/pom.xml (hunk 6 above)
Comment (mknblch) [inline tests/pom.xml]: done

File: tests/pom.xml (hunk 5 above)
Comment (mknblch) [inline tests/pom.xml]: done

File: pom.xml (hunk 4 above)
Comment (mknblch) [inline pom.xml]: done

File: pom.xml (hunk 2 above)
Comment (mknblch) [inline pom.xml]: done

File: pom.xml (hunk 3 above)
Comment (mknblch) [inline pom.xml]: done

File: algo/pom.xml [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+
+    <dependency>
+      <groupId>org.neo4j</groupId>
+      <artifactId>neo4j</artifactId>
+      <scope>provided</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.neo4j</groupId>
+      <artifactId>procedure-compiler</artifactId>
```
Comment (mknblch) [inline algo/pom.xml]: done

File: algo/pom.xml [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+  <artifactId>graph-algorithms-algo</artifactId>
+  <name>Neo4j Graph Algorithms :: Algo</name>
+  <description>Efficient Graph Algorithms for Neo4j</description>
+  <packaging>jar</packaging>
+
+  <dependencies>
+
+      <dependency>
+          <groupId>org.neo4j</groupId>
+          <artifactId>graph-algrotihms-core</artifactId>
```
Comment (mknblch) [inline algo/pom.xml]: done

File: algo/src/main/java/org/neo4j/graphalgo/impl/PageRankKernel.java [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    private final Graph graph;
+    private final double dampingFactor;
+    private final double alpha;
+
+    private double sum;
+
+    public PageRankKernel(Graph graph, double dampingFactor) {
+        this.graph = graph;
+        this.dampingFactor = dampingFactor;
+        int nodeCount = graph.vertexCount();
```
Comment (mknblch) [inline algo/src/main/java/org/neo4j/graphalgo/impl/PageRankKernel.java]: done

File: core/src/main/java/org/neo4j/graphalgo/SimpleTimer.java [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+        } else {
+            System.out.printf("%s took %dms%n", message, duration);
+        }
+    }
+
+    public long getDuration() {
+        return duration;
+    }
+
+    public static SimpleTimer start() {
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/SimpleTimer.java]: done

File: core/src/main/java/org/neo4j/graphalgo/SimpleTimer.java [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public SimpleTimer stop() {
+        duration = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
+        return this;
+    }
+
+    public void dump() {
+        if (null == message) {
+            System.out.printf("measurement took %dms%n", duration);
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/SimpleTimer.java]: done

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java [hunk 25]
```diff
@@ -0,0 +1,311 @@
...
+
+    private static class EdgeIterator implements Iterator<EdgeCursor> {
+
+        private final EdgeCursor cursor;
+        private final int[] data;
+        private final long[] ids;
+        private final int elements;
+        private int offset = 0;
+
+        private EdgeIterator(int vertexId, int[] data, long[] ids, int elements) {
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java]: done

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java [hunk 26]
```diff
@@ -0,0 +1,311 @@
...
+        this.incoming = new int[nodeCount][];
+        this.outgoingIds = new long[nodeCount][];
+        this.incomingIds = new long[nodeCount][];
+        Arrays.fill(outgoing, EMPTY_INTS);
+        Arrays.fill(incoming, EMPTY_INTS);
+        Arrays.fill(outgoingIds, EMPTY_LONGS);
+        Arrays.fill(incomingIds, EMPTY_LONGS);
+    }
+
+    AdjacencyMatrix(
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java]: one ctor is for loading with factory, the other one is used by the FileLoader.

Task: loading from file still necessary?

File: algo/src/main/java/org/neo4j/graphalgo/impl/PageRankKernel.java [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public void accept(final int node) {
+        sum = 0;
+        graph.forEachRelation(node, Direction.INCOMING, this);
+        final double value = alpha + dampingFactor * sum;
+        pageRank[node] = value;
+    }
+
+    @Override
+    public void accept(int sourceVertexId, int targetVertexId, long edgeId) {
```
Comment (jexp) [inline algo/src/main/java/org/neo4j/graphalgo/impl/PageRankKernel.java]: replace vertex/edge  in parameter names and local variables too

File: core/src/main/java/org/neo4j/graphalgo/core/WeightMapping.java [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+     * return the capacity
+     */
+    public int capacity() {
+        return capacity;
+    }
+
+    /**
+     * return primitive map for the weights
+     */
+    public LongDoubleMap weights() {
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/WeightMapping.java]: We consider the current impl. of the weight mapping as work-in-progress until we find an elegant way for storing/accessing them

**Pull request merged**